README.md
Instant Comment
===============

This script allows commenting your friends' entries without any page reloads. In place.

**<a href="https://github.com/tymofij/instant_comment/raw/master/instant_comment.user.js">Install Greasemonkey script </a>**

<img src="https://raw.githubusercontent.com/tymofij/instant_comment/master/screenshot.png" alt="example" />

Understands a bit of WackoWiki markup:

* \*\*word\*\* ⇒ **word**
* //word// ⇒ *word*
* --word-- ⇒ ~~word~~
* \_\_word\_\_ ⇒ underlined word

Creates Github-style citations from ((text))

Adds some Unicode prettifications:

* proper “quotes”. «quotes» for cyrillic texts.
* (c) ⇒ ©
* (r) ⇒ ®
* (tm) ⇒ ™
* ... ⇒ …
* uses n-dash and m-dash

Note to Mac users
------------------
You can use  Alt+Enter.



changelog
Changelog
---------

2010-12-18: **0.13**

* Added icon

2007-11-13: **0.12**

* Fixed bug with posting multiline entries appearing not sent

2007-11-11: **0.11**

* Fixed bug with posting to journals with dash in the name


2006-03-27: **0.10**

* Plays better with _Ctrl+Enter_ (update the latter too if you were using it)
* Does not ask for username
* Shows red border and alert message when posting fails


2006-03-11: **0.9**

* LJ devs added auth-challenge to comment posting form.
Therefore script now sends two requests - one to get challenge
and one to actually post message.


2006-01-24: **0.8**

* Cannot read username from cookie now.  
Therefore, you will be prompted for it first time you run script.  
Later you can change it in _about:config_, key   
_greasemonkey.scriptvals.clear.com.ua/LJ Instant Comment .username_

2006-01-19: **0.7**

* Works with new LiveJournal urls, simplified customizations.  
Just hit Edit and change constants at the top of the script.


2005-08-23: **0.6**

* Firefox 1.5 aka DeerPark compatibility

2005-07-08: **0.5**

* added alt+enter for Mac users
* proper string encoding - thanks LJ_atrustheotaku

2005-07-07: **0.4**

* fixed issue with & - thanks LJ_drdaeman
* fixed issue with &nc=
* fixed issue with lost keyboard focus
* added ability to close box with Esc key. 	
quirk - adds caret in Firefox 1.0, everything is fine in Trunk.
* added login reminder

2005-07-04: **0.3**

* fixed issue with jumping page in Firefox 1.0
* fixed some layout bugs

2005-07-02: **0.2**

* fancy progres notifier
* dynamic inputbox
* ctrl+enter

2005-06-30: **0.1**
	
* just works, uses prompt to get user input


lj_instant_comment.user.js
// ==UserScript==
// @name          LJ Instant Comment
// @description	  Adds "Instant Comment" link besides "Comment on this"
// @author        Tim Babych
// @homepage      http://clear.com.ua/projects/firefox/instant_comment
// @version       0.13
// @include       http://*.livejournal.com/*
// @icon          http://clear.com.ua/projects/instant_comment/paperplane.png
// @namespace     tymofiy_lj_instant_comment
// ==/UserScript==

// change those to customize your input box
_window_width  = 300
_window_height = 70
_left_offset = 200

// change to customize Instant Links naming
_link_title = "Instant Comment"

//================================
//	WORKHORSES
//================================

// triggerd by clicking Instant Comment link
function post_comment(){
	
	dd = document.getElementById('instant_comment')
	
	if (dd.getAttribute('caller_entry_id') == this.id) // hide-unhide
		if (dd.style.display == 'block' ) {
			hide_div(dd)
			return
		} else
			dd.style.display = 'block'
	else {
		// insert it in da new place
		dd.style.left = (findPosX(this) - _left_offset) + "px"
		dd.style.top = (findPosY(this) + 20) + "px"
		dd.setAttribute('caller_entry_id', this.id)
		dd.firstChild.value = ''
		dd.style.display = 'block'
	} 
	
	// workaround near firefox 1.0 bug with focus()
	hScroll = window.pageXOffset; vScroll = window.pageYOffset
	dd.firstChild.focus()
	window.scrollTo(hScroll, vScroll)
}

// triggered when user presses Ctrl+Enter
function trigger_submit_on_ctrl_enter(e) {
	//hide on Esc
	if (e.keyCode==27) {
		hide_div(this.parentNode)
	}

	// not enter and (ctrl or alt)
	if (! (e.keyCode==13 && (e.ctrlKey || e.altKey))) 
		return
	
	msgText = zakavych(this.value)
	msgEncoded = encodeURIComponent(msgText)

	reply = document.getElementById(this.parentNode.getAttribute('caller_entry_id'))
	reply.style.border = "3px solid rgb(0,250,0)"
	reply.style.padding = "3px"
	reply.setAttribute('mycolor', 200)
	reply.setAttribute('step', -5)

	fading = function(){
		elem = reply
		mycolor = parseInt(elem.getAttribute('mycolor'))
		step = parseInt(elem.getAttribute('step'))
		if ( mycolor> 250)
			step -= 5
		if ( mycolor < 50)
			step += 5
		mycolor += step
		elem.style.borderColor = "rgb("+mycolor+", 250,"+mycolor+")"
		elem.setAttribute('mycolor', mycolor)
		reply.setAttribute('step', step)
	}

	anim = window.setInterval(fading, 15)
	hide_div(this.parentNode)
	
	GM_xmlhttpRequest({
	    method: 'GET',
	    url: reply.getAttribute('orig_href'),
	    onload: function(responseDetails) {
		var s = responseDetails.responseText
		try {
		    tokens = ["lj_form_auth", "chal", "cookieuser", "chrp1", "journal"]
		    token_str = '&usertype=cookieuser'
		    for (key in tokens) {
		        token = tokens[key]
		        r = new RegExp('["\']'+token+'["\'].+?value=["\'](\\S+?)["\']')
		        token_str += '&'+token+'='+r.exec(s)[1]
		    }
		} catch(err) {
			window.clearInterval(anim); 
			reply.style.borderColor="red"			
			window.alert('Error getting LJ auth token. \nMake sure you are logged in and not banned.')
			return
		}
		GM_xmlhttpRequest({
			method: 'POST',
			url: 'http://www.livejournal.com/talkpost_do.bml',
			headers: {
				'User-agent': 'Mozilla/4.0 (compatible) Greasemonkey Instant Comment',
				'Content-type': 'application/x-www-form-urlencoded',
			},
    		data: 'itemid='+reply.getAttribute('itemid')+'&body='+msgEncoded+token_str,
			onload: function(responseDetails) {
				window.clearInterval(anim); 
				// we check for the first line of the comment				
				needle = msgText.split("\n")[0]
				// or just for URL in that line
				if (urls_in_post = needle.match('https?://\\S+'))
				    needle = urls_in_post[0]
				    
				if (responseDetails.responseText.indexOf(needle) == -1) {
				    // Text was not found in responce
					reply.style.borderColor="red"
					homepage = "\n\nIf the error repeats, please contact me.\nTim Babych, http://clear.com.ua"
					if (responseDetails.responseText.indexOf("Frank's nibbling on the wires") != -1) {
					    window.alert("Your comment was not posted. \nFrank's nibbling on the wires."+homepage)
					} else {
            			window.alert('Your comment was not posted.\nPosted msg was "'+msgText+'"'+homepage)
            		
            		// UNCOMMENT This to see actual error page given by LJ	
                    // window.document.body.innerHTML = responseDetails.responseText
            		}
				} else {
					reply.style.borderColor="green"				
				}
			}
		});

		
	    }
	})
}

// hide input div, return focus to page
function hide_div(div){
	div.style.display = 'none'
	div.firstChild.blur()
}


//================================
//	INIT
//================================

d = document.createElement('DIV')
d.id = "instant_comment"
d.setAttribute('caller_entry_id', false)
if (navigator.userAgent.indexOf("Mac") !=-1)
    hint = 'Alt+Enter to post'
else
    hint = 'Ctrl+Enter to post'
d.innerHTML = "<textarea id='instant_comment_textarea'></textarea><br/><small>"+hint+"</small>"
d.firstChild.addEventListener("keydown", trigger_submit_on_ctrl_enter, false);
document.body.appendChild(d)
//--------------------------------------------------------

/* 
NAME.livejournal.com/123456.html
community.livejournal.com/NAME/123456.html
users.livejournal.com/NAME/123456.html
*/
get_itemid_regexp = /([\w\d-]+)\.livejournal\.com\/([\w\d]*\/)?(\d+)\.html/ 
var allReplies
allReplies = document.evaluate(
    "//a[contains(@href, '?mode=reply')]",
    document,
    null,
    XPathResult.UNORDERED_NODE_SNAPSHOT_TYPE,
    null);
for (var i = 0; i < allReplies.snapshotLength; i++) {
	thisOne = allReplies.snapshotItem(i);

	t = document.createTextNode(' - ')
	thisOne.parentNode.insertBefore(t, thisOne.nextSibling);
	
	params =  get_itemid_regexp.exec(thisOne.href)

	if (!params)
		// unmatched url, may be dead or greatest journal ?
		continue

	a = document.createElement('a')
	linktxt = document.createTextNode(_link_title)
	a.href = 'javascript:void(0)'
	
	if (params[1] == 'community' || params[1] == 'users')
		journal =  params[2].substr(0, params[2].length -1)
	else 
		journal = params[1]
		
	itemid = params[3]

	a.setAttribute('itemid', itemid)
	a.setAttribute('orig_href', thisOne.href)
	a.id = 'insta_'+journal + itemid
	a.addEventListener("click", post_comment, false);	
	a.appendChild(linktxt)
	t.parentNode.insertBefore(a, t.nextSibling);
}
//-----------------------------------------------------------------------
GM_addStyle(
" #instant_comment { position: absolute; display: none;	border: 1px solid #999;"+
"	background: #ececec; padding: 1px; text-align: center; z-index:99}"+
"#instant_comment textarea { width:"+_window_width+"px; height:"+_window_height+"px; min-height: 10px;"+
"	margin: 1px; border:1px solid #999; padding-left: 3px;	background: white; color:black; }"+
"#instant_comment textarea:focus { border:1px solid black; }"+
"#instant_comment small { color:#999; font: 10px Arial }"
)
//-----------------------------------------------------------------------

//====================================
//	ROUTINES
//====================================

function findPosX(obj) {
	var curleft = 0;
	if (obj.offsetParent) {
		while (obj.offsetParent) {
			curleft += obj.offsetLeft
			obj = obj.offsetParent;
		}
	} 
	return curleft;
}

function findPosY(obj) {
	var curtop = 0;
	if (obj.offsetParent) {
		while (obj.offsetParent) {
			curtop += obj.offsetTop
			obj = obj.offsetParent;
		}
	} 
	return curtop;
}

function zakavych(text) {
	
    replacements = [
    
    // latin text in english quotes
    [/(\s+|^)"([^\"а-ягўєїА-ЯҐЎЄЇ]+?)"(\s+|$|\.|\,|\!|\?)/g, '$1“$2”$3'],
    
    // other text in lapky
    [/(\s+|^)"([^\"]+?)"(\s+|$|\.|\,|\!|\?)/g, '$1«$2»$3'],

   
    // ukrainian apostrophe
    [/([б-щБ-ЩҐ])[\*'`]([а-яєїА-ЯЄЇ])/g, '$1’$2'],
    
    // trademark (TM) and such
    [/\((tm|TM|тм|ТМ)\)/g, '™'],
    
    // copyright (C) and such
    [/\([cCсС]\)/g, '©'],    
    
    // registered (R) and such
    [/\([rRрР]\)/g, '\®'],
    
    // mdash -- one or two minuses surrounded by spaces
    [/(\s+|^)--?(\s+)/g, '$1\u—$2'],
    
    // **bold**	
    [/\*{2}([^\*]+?)\*{2}/g, '<b>$1</b>'],
    
    // //italic//
    [/([^\:]|^)\/{2}(.+?[^:])\/{2}/g, '$1<i>$2</i>'],
    
    // --strikeout--
    [/([^\!]|^)-{2}([^-]+?)-{2}/g, '$1<s>$2</s>'],

    // __underlined__
    [/_{2}([^_]+?)_{2}/g, '<u>$1</u>'],
    
    // ndash for number ranges: 1995-2005
    [/(\s)(\d+)-(\d+)(\s)/g, '$1$2–$3$4'],
    
    // ellipsis	
    [/\.\.\./g, '…'],
    
    // strip extra LFs at the end
    [/\n*$/, '']
    ];

	s = text
	for( i=0; i < replacements.length; i++) {
		s = s.replace(replacements[i][0], replacements[i][1])
	}
	
	return s
}
